### DISCUSSION TITLE
Juggluco + Patched .apk Q&A -> Post your questions here

### DISCUSSION THREAD

**Discussion by maheini** [maheini/FreeStyle-Libre-3-patch#12] [Q&A]
### Which phones are supported?

Currently, there are 64bit phones supported which are not rooted

### The libre app is crashing randomly...

Check if your phone has root access. In order to use the app, you have to unroot it or hide your root. Here's a instruction on how to hide the root: [How to hide root](https://www.reddit.com/r/Freestylelibre/comments/s22vlr/comment/hw2p4th/?utm_source=share&utm_medium=web2x&context=3)

### I don't get a connection between libre 3 and juggluco

Check the instructions again and read carefully. If you still face connection issues, then check if you missed any checkbock within libre 3 (poorly visible on the red background).

### The patched .apk states it only supports sensors from Germany

That's no problem. Just keep going and ignore the warning - you should be able to activate any kind of libre 3 sensors.

### Is the patch trustworthy?

AFAIK, jkaltes didn't released the source code of the patch. In general the patch is trustworthy, but there's no guarantee - and never have been for Apps like patches, AAPS, Xdrip etc. It's on your own risk.
Just let me assure you that there is a contact option for jkaltes. If you got an urgent request to jkaltes, feel free to contact me. As a second guarantee: Juggluco is open-source, giving some sort of security. Still there's no guarantee regarding the functionality, as in every other app as well.

### I'm having error-code "postgetauth: status=10 reason=invalidUserOrPassword"

See this answer: [Link](https://github.com/maheini/FreeStyle-Libre-3-patch/discussions/12#discussioncomment-4945156)

### There are often no readings after the sensor signal got lost

The problem might be related to Android 13: [link](https://github.com/maheini/FreeStyle-Libre-3-patch/discussions/12#discussioncomment-4756657)

**Comment by dsnomr** [comment]
What do I have to do to take over an already started sensor with the patched L3 app or to restart it in the patched app? I am told that the sensor is already in use and therefore cannot be used with the patched app.

**Reply by maheini** [reply]
As far as I know, you shoul be able to reconnect a already running sensor. Do you use the same libre account on both apps? And do you still use the original libre 3 app at the same time?

> On one mobile phone, the started sensor is running with the original L3 app, I would like to transfer this to another mobile phone with the patched app, the same libre account is used on both.

Well then... I got you a new attempt: 

**On your current phone with original libre3 app**
- Shut down your current phone (to disable the libre3 connection). 

**On your second phone with the patched version**
- force stop the patched app via settings->app...
- now start the patched version again and tap on "Start new Sensor", as if you start a (really) new sensor. Now you should get into the warmup period of the sensor (even if the sensor is new) and libre 3 should receive readings after 1 hour.

Just one last question: Where you signed in using your Libre account, at the moment you activated the libre3 sensor for the first time? This might be a problem as well.

**Comment by dsnomr** [comment]
I have already tried it several times without success. The sensor always remains connected to the first mobile phone.
It could be that I was not logged in with my account when I started the L3 sensor with the original L3 app. A subsequent login was possible there. That is probably the problem.
Many thanks for your effort and the explanation!

**Comment by Nb-tst** [comment]
Using libre3 ( patched  ) + juggluco + xdrip
So far everything is working ok, connection is more stable than with FS2+OOP2, no disconnections when in range, istant reconnections when you return in range, working at a bigger range from phone.
I wanna ask a question about sensor's expiration, does it last exactly 14 days or as FS2 there is a 12h bonus?

**Reply by maheini** [reply]
Hi 😄 
Great to hear that everything is working perfect and stable. 
Regarding the extra 12 hours extra:
I'd expect it to still be possible, because it's technically almost impossible to prevent that (It's made for traveling into different time zones). However, I don't know for sure and would be very glad if you could post any further news (If you actually try to extend the sensor by changing your system time).

If you got any other questions, feel free to ask.

**Reply by twinko** [reply]
can u guys explane how to pull of the extra 12h?
And i never understood why people use libre2 and oop, i just used l2 and patched app. Regards :)

**Reply by Nb-tst** [reply]
IDK, with oop2 by default it last 14.5 days. Which is very useful, as I can easily reach 15 days between two sensors activation adding a night.

**Reply by Nb-tst** [reply]
Update: with my configuration ( patched libre3 + juggluco + xdrip ) the sensor last exactly 14 days, so no 12h bonus as for FLS2.
As this configuration relies primarily on libre3 app for data transmission with the sensor, I cannot stop the app in order not to send the shutdown message.

**Comment by gui-dos** [comment]


**Reply by Nb-tst** [reply]
So, there could be a way to patch libre3 app in order to make the sensor last 15 days.

**Reply by maheini** [reply]
Hi @gui-dos 
I saw your messages - thank you 😄. I'm able to contact jkaltes - should I send him anything you mentioned? If you'd like to contact him directly, he's a member of insulinclub.de (as you are too). He's currently trying to extend the sensor as well, but not yet successful.
Feel free to reply at any time - thanks.

**Reply by MasterPlexus** [reply]
To explain that more further. The Libre 2 app could be patched hat the ramp-up time will be only 45 minutes, and that the living time of the Sensor could be extended by 12 hours. This was never "official in the patches" as there is an extension in the lifetime, wich could may produce may wrong values, so if this was needed the patch must be done by known people by themselves (anyhow the patched app was also to be done by enyone normaly themself ;) ). But of course I don't know if the times was now more stringed in the hardware of L3 now itself.

**Comment by TarcisoSilva** [comment]
Using libre3 ( patched ) + juggluco + xdrip + APS
So far everything is working ok. No isses at this moment. Connection is OK.

**Reply by ezepizarro** [reply]
Hi, I am trying to the same but I found out that I am only getting new glucose values every 5 minutes and not 1 minute intervals.
Are you getting new readings every 1 minute?

**Reply by maheini** [reply]
Hi @ezepizarro 
It's totally normal to receive only a value every 5 minutes within xdrip. This value is calculated by xdrip itself and based of the last 20 minutes of libre2 readings. 
Check Settings > Less common settings > Advanced settings for Libre 2 > show Raw values in Graph, then you'll be able to see all the readings within xdrip as white dots.

**Reply by ezepizarro** [reply]
@maheini What about receiving the values every 1 minute as the Libre 3 is capable of doing that? 
Using the libre 3 app bridge, I can also have 1-minute readings in my NS and AAPS. I was wondering if we can do the same with xDrip.

**Reply by maheini** [reply]
It's possible to get a calculated reading every minute... however, you'll need to modify the source code and build xDrip yourself. As far as I know, there's no ready-to-use solution. 

Btw. whats "libre 3 app bridge"? I don't know that... may you send me a quick link?

**Reply by ezepizarro** [reply]
@maheini You already know about the libre 3 bridge, it's the libre link uploader. I don't know why but I wrote"bridge".
https://github.com/timoschlueter/nightscout-librelink-up

Do you know where to modify the xDrip code to get the readings every minute?

**Reply by ezepizarro** [reply]
I found this xDrip that gets the data from NS every 1 minute.

https://github.com/jakobheyman/xdripJH

**Comment by TarcisoSilva** [comment]
Just a feedback. 
When ADD a CONNECTION on the instructions page the port informed is 7117, the correct should be 8795. I belive there is a typo there. Can you please confirm it?

**Reply by maheini** [reply]
Hi @TarcisoSilva 
Thanks for your reply. Which step in particular do you mean? As far as I know, all port setting should be correctly, as jkaltes states in his instructions. What ports did you set on libre3 "juggluco page" & libre3 "connection page" and which one in "juggluco mirror" & juggluco "mirror connection settings"?

Both ports should be used within juggluco and libre3 - unfortunately, they are interchanged within those two apps.

**Reply by TarcisoSilva** [reply]
![image](https://user-images.githubusercontent.com/86031348/195141058-c35156d7-441c-4506-8673-db19620f1b31.jpeg)

**Reply by TarcisoSilva** [reply]
Please take a look on picture. The port used should be 8795 instead 7117, correct? 
I am using 8795 on my connection now and it is working. I tried 7117 with no success

**Reply by maheini** [reply]
Hmm... I wasn't aware of that till now. Thanks!

Just one last question: Are you using the port "7117" anywhere at all? (libre3 app or juggluco app?)

I'm kind of confused... can't try it myself (as I don't have access to libre 3 sensors) and all instructions online are stating the same as mine do...

**Reply by TarcisoSilva** [reply]
No, I am not using port 7117

**Reply by TarcisoSilva** [reply]
This is my configuration now![image](https://user-images.githubusercontent.com/86031348/195167376-244cbf1d-2a6e-48b9-8873-cb36928f1fa6.jpeg)

**Reply by maheini** [reply]
So, you are not using the port 7117, even in the "libre3 app -> juggluco" settings? That might be the "cause" then. In the "official" instructions, it's stated like here - with two different ports.

**Reply by TarcisoSilva** [reply]
I am not using 7117. ![image](https://user-images.githubusercontent.com/86031348/195197698-dc3c393e-ccf8-4ece-bac8-6cd5b91e77b4.jpeg)

**Reply by sheh1000** [reply]
I'm sure the instruction has no typo in port settings. 
So you have 2 Juggluco instances: first one is integrated into Librelink app, and dedicated Juggluco app.
When you create new connection, first port means what port is listening on for current app.
Second port (after you press "Add connection" button) means destination port where you are connection to.
So in the end you should have cross-connection between 2 Juggluco instances:

```
First screen             | Add connection screen
(listen on)              | (send to)
---------------------------------------------------
Librelink Juggluco:7117  | Juggluco:8795
Juggluco:8795            | LibreLink Juggluco:7117
```

**Reply by maheini** [reply]
@sheh1000 Thank you so much for clearing things up. Although I knew it's probably wrong, I still got quite confused - because I don't had any possibility to test it myself.

**Comment by sheh1000** [comment]
I'm having issues with getting data from Juggluco to Xdrip+. My end goal is to forward Libre 3 sensor data to Nightscout. I can see my sensor data from Juggluco, but not from Xdrip+
I carefully followed [steps 4 and 5](https://github.com/maheini/FreeStyle-Libre-3-patch/blob/main/Juggluco-solution/instructions/en/instructions.md#step-4-optional-set-up-xdrip) but few moments not clear to me:

1. How xdrip receiving data from Juggluco? Is it listening on some port? Here is the quote from [code comments](https://github.com/NightscoutFoundation/xDrip/blob/424f32a37e74c98c4af150eaa8ad774c669b913f/app/src/main/java/com/eveningoutpost/dexdrip/webservices/XdripWebService.java#L47) that xDrip+ runs a webservice on localhost port 17580 respond to incoming requests either for data or to push events in
2. should I create a second mirror connection from Juggluco for sending data to XDrip+ to `127.0.0.1:17580`?
3. is there any way to emulate recieving sensor data to Juggluco? Could be helpful for troubleshooting without active sensor  
4. Could Juggluco mirror data to [Glimp](https://play.google.com/store/apps/details?id=it.ct.glicemia&hl=ru&gl=GB) or any other app that could report data to Nightscout?

**Reply by maheini** [reply]
Hi @sheh1000 

1. Xdrip is receiving the values with the Data source "Libre 2 (patched App)". I don't know a lot about this source,  but I think it's very stable.
2. No. The only thing required is to enable "Send to xDrip" within the Juggluco settings.
3. If you still don't get any readings, the problem is most likely caused by xDrip+ or Juggluco. If you receive any reading within Juggluco or Libre3 app, the sensor should be fine.
4. Not at this moment (at least not without xdrip. If you are dependent on Nightscout, things like "LLU client" might be an alternative for you. You can find informations about this solutions within the readme file in this repository.

Feel free to ask any remaining questions and have a nice week :)

**Comment by Jonaskjellsson** [comment]
Is it possible to go directly from patched.apk L3 to Aaps?

Or directly from juggluco to Aaps?

**Reply by maheini** [reply]
@Jonaskjellsson 
Currently, there's no option to receive the readings directly within AAPS or Xdrip+. I heard of some work ongoing within xdrip+, to receive the juggluco readings directly. But I don't know anything more as of now.

However it's not that hard to set up xdrip...

**Reply by twinko** [reply]
> I heard of some work ongoing within xdrip+, to receive the juggluco readings directly. 

Do you mean "to receive *patched libre 3 app* readings directly." ?

**Reply by maheini** [reply]
I think so. But I just caught that in passing.

**Reply by twinko** [reply]
would be really interested where to read about this development :)

**Reply by maheini** [reply]
Hmmm... I'm trying to find it - but didn't even know myself where I got those informations from. I will keep you updated 😄

**Comment by sheh1000** [comment]
My sensor has lost a connection with patched LibreLink app. I've been waiting for almost 40 minutes, even tried to reboot my phone and still the same. What happens if I execute "Start New Sensor" button in LibreLink, but scan the same sensor that got disconnected? Is it safe for sensor to scan it again?

**Reply by maheini** [reply]
If your sensor lost connection to the patched libre 3 app, it's most likely not a caused by the patch. As far as I know, you won't be able to restart an existing sensor again if he's already started within your current app (nothing should change or happen). It shouldn't be a problem at least - but it won't help you as well.

Did you signed in with your libre account before you started your sensor? If so, you can just try to download the original Libre3 App and use it like this.

**Reply by sheh1000** [reply]
If anyone has the same issue - I turned off Bluetooth on the old phone where previous sensor was active with patched LibreLink app. Probably, it somehow interacted with Bluetooth connection on my current phone. Anyway, turning off BT connection on another phone helped.

**Reply by AgushkaNightscout** [reply]
I had not the same issue, but similar. I use a fsl3 sensor with juggluco -> AAPS. At some point the 10 day old sensor with a near expiration date showed an "sensor error", then "no connection" and the sensor stopped working. It was responding to the juggluco scan, but it was not transmitting data. I installed a new one (and started with patched fsl3 app), but I did not remove the old one and it miraculously came back to life after 2 (!!!) hours. There were 2 curves in the program and from that moment into AAPS automatically started transmitting information from the old sensor. I did not measure exactly how long it was, but the next day the readings from the juggluco were not from one, but from both sensors alternately, so I terminate old sensor.

**Comment by Nb-tst** [comment]
jka has updated his app, now juggluco does not need librelink app in order to connect with FSL3. I don't know if this could let the sensor last more than 14 days ( I don't speak german so I cannot ask him in that forum ).
https://insulinclub.de/index.php?thread/34255-libre3-2-juggluco/&postID=669283#post669283

**Reply by maheini** [reply]
Hi @Nb-tst 
Thanks for the update - I already heard about that. I planned to update the instructions, but it will probably take some time (The solution & instructions need to be bulletproof & tested).

The new version doesn't seem to extend the sensor life as far as I see. But jkaltes didn't mentioned the sensor duration at all.

**Reply by Nb-tst** [reply]
I have a sensor ending soon, I could give a try. the last day.
 IDK if I will be able to switch from FSL3 to juggluco with an active sensor, in case I will discover if even jugglugo sends the termination code to the sensor.

**Reply by twinko** [reply]
Translatetd part from jka's post regarding running sensors:
To use your current sensor, you need to specify your LibreView account under Settings -> Send to Libreview and click on "Get Account ID" and wait. Later you can check if it was received from the number displayed before "Get Account ID". After that you can scan the sensor.

**Reply by maheini** [reply]
Oh, I would really apprechiate if, if your're able to try this new solution and test whether it's extending the sensor. Just post your results here and I will update the instructions.

**Reply by Nb-tst** [reply]
I never connect my sensor to libreview, so I think that the first thing is to create an account and try to connect it.

**Comment by Nb-tst** [comment]
I tried to  to connect FSL3 to Juggluco, following jka's instruction. I got the ID from libreview but nonetheless everytime ( except once ) I scanned the sensor with NFC Juggluco gave me the "wrong account ID?" error. Once it gave me a different answer, but it was anyway unable to find any bluetooth device during its scanning.
 After that  the sensor gave me some errors when I opened again the FS app, and I had to wait nearly an hour to establish again a working connection.
 So, this time I was not able to try to extend sensor's lifetime. I saw that meawhile Juggluco has been updated many times, from 3.0.0 to 3.0.4, so maybe with the next sensor it will work.

**Reply by j-kaltes** [reply]
Sound like you used the wrong account email address. That it worked one time sounds implausible. 
Juggluco can't find Freestyle Libre 3 sensors by Bluetooth scanning. After NFC scanning the device address of the sensor is known. Under Left menu->Sensor you can see it. Scanning adds nothing.
Did you see the sensor under Left menu->Sensor? 
I can quickly switch between apps without much delay finding a connection. It is important to kill the competing app and to turn Bluetooth off and on.
Did you have root? Juggluco will crash and start again if it is not hidden.
If you use a new version of the patched Libre3 app, you can also get Juggluco (if it is one that supports Libre3), to which it sends data, to take over the connection without scanning. Just force close the Libre3 app, reverse or remove the Mirror connection and turn on "Sensor via Bluetooth".

**Reply by Nb-tst** [reply]
Thanks for your reply and your precious work.
- No root and locked bootloader.
- email and password verified multiple times
- I killed librelink3 but I did not switch on and off BT ( or maybe I tried sometime, I've tried a lot of times with minor differences in the procedure in order to find a right one )
- Under "sensor" all fields were and remained empty, tried to press reenable, tried to check turbo, but nothing happened.
- I did not know that there was a new FSL3 patched app, I will search for it.
- I saw that from 3.0.2 the app size is one third compared to previous versions, have you removed something?

**Reply by j-kaltes** [reply]
Reenable is useless for Libre3 and you always should rescan (NFC) thereafter.
Maybe you downloaded the arm64 only version (Size: 14.65 MB); that is 37% the size of the version with arm/arm64/x86/x86_64 (Size: 39.47 MB).

**Reply by Nb-tst** [reply]
Yes, I downloaded that one. I will try again with the new sensor and I will report results.

**Reply by Nb-tst** [reply]
Same issue with the new sensor, not able to to connect juggluco, i got the very same error when I scan with nfc. In the sensors screen in juggluco, it appears a sensor code  but at the same time it displays an "address unknown" and a "streaming not enabled" status.
Maybe it's because I logged in freelink app with my credentials only after I started the sensor or I am missing some basic setting.

**Reply by j-kaltes** [reply]
> Maybe it's because I logged in freelink app with my credentials only after I started the sensor or I am missing some basic setting.

You have to give to Juggluco the Libreview Account data you used while starting the sensor. An account id number is send to the sensor during the NFC activation and later the sensor refuses to function when it is not the same. If you don't use an Libreview account in the Libre3 app a random number is generated and used.

**Comment by AgushkaNightscout** [comment]
Hello. Big thanks to Jkaltes for his brilliant app! 
2 weeks ago I connected libre3DE sensor to xdrip+ via bundle of patched libre3de app and patched juggluco and had no problems with stability of connection. But yesterday I connected new libre3DE sensor via only patched juggluco 3.0.4 arm-only version and had 2 different issues. First one (I suppose, when sensor was pressed down) - no readings for almost 40 minutes with status 9 (settings -> sensor) and second one - 2 times no readings for 10 minutes with status 133 (maybe because of the problems of using bluetooth by xdrip-latest aaps-3.1.0.2 and juggluco at the same time). I switched off bt watchdogs in xdpir and aaps and the problem has not occurred for 24 hours, but that's not really right, is it? Could you tell me please what is the right thing to do in this case?

**Reply by AgushkaNightscout** [reply]
Two stops just happened - status 133 and new status 19, both for 15 minutes. Turned on AAPS bluetooth watchdog.

**Reply by j-kaltes** [reply]
When Juggluco is connected with a Freestyle Libre 3 sensor and I start the Libre 3 app and scan the sensor with it like it was a new sensor, I receive in Juggluco a connection error with status=19.
The solution is to kill the Libre 3 app and scan the sensor again with Juggluco. It will never regain a working connection with the sensor without re-scanning the sensor.

**Reply by AgushkaNightscout** [reply]
Thank you, J-kaltes for your answer! You are doing very important job for every d1t people around the world! 
I killed original Libre 3DE app yesterday and didin't run it after. Yes, I scan the sensor with Juggluco every time when readings disappears.

**Reply by AgushkaNightscout** [reply]
After a week of using librе3 sensor directly via Juggluco, one pattern has emerged - if the bg is rising fast (e.g. after a meal), then after 10 mmol the connection to the sensor is lost with a status=19 (mostly) and appears when the reading goes down, also from about 10 mmol.

**Reply by j-kaltes** [reply]
onConnectionStateChange with status=19 should be called by Android when the device (the sensor) ends the connection. 
https://medium.com/@martijn.van.welie/making-android-ble-work-part-2-47a3cdaade07
It happens for example when you scan the sensor. 
Also when the sensor is not scanned it now and then happens, but in my experience it does not often make that no value is available.
In my the sparse experience with status=19, I didn't see any relation with the glucose curve and don't see why they should program the sensor that way.
EDIT:
Just now I had this status=19 on my watch; causing more than 3 minutes without new value. I see more a relation between this error and me not looking at my watch. There was immediately a new value when I switch to Juggluco on my watch.

**Reply by AgushkaNightscout** [reply]
Thank you for the link with the description of the codes! 
Clarification - the first break occurs with code 133 or 9. When I scan the sensor all next codes are 19. Perhaps an additional factor in the breakup is not just the bg value, but the rate of change. I have looked at the history again and all the gaps were at values over 10mmol and speed over 1.5mmol. There were no such rates or values today, nor were there any discontinuities. Next sensor I will activate with patched libre3 app and see how it works with patched Juggluco.

**Reply by j-kaltes** [reply]
Gaps in the past are not connection errors, but sensor errors. Freestyle Libre 3 sensor have back fill by which missed values are send over again. You don't see connection errors of the past. If the sensor error lasts long enough, also the history values are disrupted. When you use Abbott's Libre 3 app and download the values from LibreView, you will also see the same gaps.
Sensor errors happen especially after abrupt changes.
EDIT:
On Juggluco 3.0.0 you can infer a **sensor erro**r from an error after Glucose Value in left menu-> Sensor. In Juggluco 4.0.1 it is displayed in place of the glucose value at the right of the curve.

There is a way find **connection errors** of the past, because the backfill values are without direction of change. Save the stream glucose values in Juggluco with Left middle menu->Export->Stream.
Values received via backfill have a rate of change of  +nan  and a change label of NOT_DETERMINED.

**Reply by AgushkaNightscout** [reply]
Do sensor errors occur because the sensor is faulty or does pinching, for example, cause this error?

I have saved the stream and at the points on the graph where there were omissions - in the stream table there are just no values, i.e. 5538 is followed by 5550 etc. I also have NOT_DETERMINED errors but I don't see them on the graph (because they are backfilled) and they are not dangerous in general. What can it be?

Could it be that I have version 3.4.0 - the latest as of 02.11.2022?
![Screenshot_20221108-220517](https://user-images.githubusercontent.com/90107639/200688206-f2bdd140-c967-45f5-85a7-4f5ff24e0761.png)
![Screenshot_20221108-220527](https://user-images.githubusercontent.com/90107639/200688210-e2ef17ac-8641-48de-8ac8-5a2c4a36e930.png)
![Screenshot_20221108-220555](https://user-images.githubusercontent.com/90107639/200688214-8d062b64-d53d-48e7-b9ab-bd2a403b560b.png)
![Screenshot_20221108-220610](https://user-images.githubusercontent.com/90107639/200688220-69c6de05-1653-4441-9073-096b8984db7c.png)
![Screenshot_20221108-220620](https://user-images.githubusercontent.com/90107639/200688225-78318c08-d902-4e52-bd48-74418243ffc5.png)
[Stream2022-11-08-22_09_34.xls](https://github.com/maheini/FreeStyle-Libre-3-patch/files/9965694/Stream2022-11-08-22_09_34.xls)

**Reply by j-kaltes** [reply]
> Could it be that I have version 3.4.0 - the latest as of 02.11.2022?

I don't understand were you are talking about? The Libre3 version? Juggluco has no version 3.4.0. Juggluco 3.0.4 should not have these problems.

You can control if these values are present in LibreView by force closing Juggluco, starting the Libre3 app, press "Start New Sensor", "Yes", "Next" and scan the sensor (something which was said to be impossible in this forum) and wait for the values to be received and send to Libreview. Then you can download the data and look for the same values there. If they are not there, you can be certain that it is a sensor error.

**Reply by twinko** [reply]
http://jkaltes.byethost16.com/Juggluco/download.html

There you'll find the latest juggluco version (4.0.1)

**Reply by AgushkaNightscout** [reply]
> I don't understand were you are talking about? The Libre3 version? Juggluco has no version 3.4.0. Juggluco 3.0.4 should not have these problems.

my mistake, of course Juggluco 3.0.4, and printscreens with .xls file also form 3.0.4

> You can control if these values are present in LibreView by force closing Juggluco, starting the Libre3 app, press "Start New Sensor", "Yes", "Next" and scan the sensor (something which was said to be impossible in this forum) and wait for the values to be received and send to Libreview. Then you can download the data and look for the same values there. If they are not there, you can be certain that it is a sensor error.

Thank you, I'll try it when sensor will be 13 days old, just to be safe. 


> There you'll find the latest juggluco version (4.0.1)

yes, I know, I downloaded from there my current version and upgrade next time

**Comment by gui-dos** [comment]


**Reply by j-kaltes** [reply]
I think that it is easier to get rid of the native libraries of the Libre3 app, than of those of Librelink 2.5.3, because the first only uses them for decryption, encryption and signing, the latter also interprets the results. Easiest of the puzzles to solve I mentioned already:
https://insulinclub.de/index.php?thread/33795-free-three-ein-xposed-lsposed-modul-f%C3%BCr-libre-3-aktueller-wert-am-sperrbildschir/&postID=655055#post655055
Finding out what checksum or crc16 they used.
Do you know how you can get the Iphone version of Abbotts apps for FreeStyle Libre 2 and 3?
And are there good de-compilation tools for iphone apps?
Isn't it possible to somehow link to an Arm linux library on Iphone when emulating a few syscalls?

**Reply by j-kaltes** [reply]
Strings on the shared libraries of the Android version shows absolutely nothing. Also the machine code they are using, you can only see when the program is running.

**Reply by j-kaltes** [reply]
The exact calculation I gave is wrong. 
This is the C++ version I use:
```
static jlong getlibreaccountidnumber() {
         auto &accountid=settings->data()->libreviewAccountID;
         uint32_t res=0;        
         for(auto el:accountid) {
            res = (res * 0x811C9DC5)^el;
            }
         const jlong lres= static_cast<jlong>(res);
         return lres;
         }

```

**Comment by Nb-tst** [comment]
I'm reporting two sensors _brickings_ EDIT _stopping to communicate values in any way to L3 App or juggluco_ trying to use juggluco ( ver 4.0.1)   direct method. The first one happened in a expiring sensor, the second one in a brand new
After trying to establish a connection scanning with nfc, as in instructions by @j-kaltes, I got "wrong account id" on the old sensor and some other "nfc" error on the new one. After that, libre3 displays a permanent signal loss status. Nfc scanning with L3 does not solve the issue ( nor can get any value ). ( this time I was logged in )
If I try to scan the sensor with juggluco it gave me a "freestyle libre 3 sensor error" yellow message.

**Reply by j-kaltes** [reply]
Is the "Wrong account id" sensor the same one you talked about earlier. You don't want to tell me that you couldn't use that sensor thereafter with the Libre3 app by scanning the sensor like it is a new sensor, will you? So it was not bricked. You only couldn't use it with Juggluco. You created a Libreview account after you had activated the sensor. That will never work. A number received from Libreview is send to the sensor during activation and that has to be the same the next time the sensor is scanned. Without Libreview account a random number is send, which you can't get to known without root access.

What do you mean scanning the sensor with Libre3? Do you use "Start new sensor" and the sensor is accepted? It makes a difference whether Libre3 app accepted the sensor or not. After every switch you should scan again like it is a new sensor.

Of overall importance is whether the Libreview Account you are using while activating the sensor is the same as while scanning the sensor, can you be clearer about this. Did you activate the sensor with Juggluco or the Libre 3 app?
EDIT:
Earlier in this forum you said: "_I never connect my sensor to libreview, so I think that the first thing is to create an account and try to connect it._".
When I am using the Libre3 app with a Libreview account and I logout the Libre3 out of that account, I can't scan the sensor anymore. I have to login again to be able to scan the sensor again. 
So you added an Libreview account to Abbott's Libre3 app. Because of this the app didn't work anymore with the sensor.
You are blaming something that has to do with the Libre3 app on Juggluco.

**Reply by Nb-tst** [reply]
It's not my intention to blame anything or anyone, I was just reporting what happened to me.
 I've managed to re establish a connection using your suggestion of starting a new sensor.  A few days ago I tried the same procedure with the old sensor, but I didn't work, so I did not even try the same with the new one until you wrote your reply. If I understand correctly, the crucial point is that I did not link the old one to libreview from its start, while I did with this new one so I can reconnect it by starting it as new when it fails. 
 Still don't understand why when I try a direct connection with Juggluco  it fails and puts the sensor in a not "reachable" mode. It can't be a wrong account/credential issue, as I have only one account and I'm using the very same credentials that worked to restart the sensor with L3 app.

**Reply by j-kaltes** [reply]
As I understand you have now the sensor started in Libre3 with a Libreview account.
In Juggluco you should specify the same account data and press "Get Account ID", after sometime a number will be displayed before "Get Account ID". You now scan the sensor with Juggluco. Sometimes you receive an error message, like "FreeStyle Libre 3 sensor. Error" or "Unrecognized NFC tag". In that case you just scan again, until you received the message "Sensor ready for use".
You should have killed the Libre3 app. "Sensor via Bluetooth" in settings ("Use Bluetooth" in left menu->Sensor) should be turned on. Normally you have to wait three minutes before Juggluco has established contact with the sensor. Sometimes turning off other apps or devices or turning Bluetooth off and on helps.
If Juggluco doesn't establish contact with the sensor in 10 minutes, send a screenshot of Left menu->Sensor.
EDIT:
Every time you scan the sensor a new pin code is send over that is needed to connect via Bluetooth. The previous pin is no longer valid. That is the reason why you need to re-scan the sensor when you switch between the Libre3 app and Juggluco.

**Reply by Nb-tst** [reply]
Thanks, I will try again in the coming days and report.

**Reply by Nb-tst** [reply]
Tested working with ver 4.19. It needed an app data/cache clean and a few nfc scans in order to start a connection.
Unfortunately, no 12h bonus after standard expiration as in FSL2.

**Reply by j-kaltes** [reply]
You are right. Today for the first time I reached the end of a Freestyle Libre 3 sensor with Juggluco and I had a 12 minutes bonus instead of the hoped for 12 hour bonus.

**Reply by j-kaltes** [reply]
I receive LIBRE3_CHAR_EVENT_LOG, LIBRE3_CHAR_CLINICAL_DATA and LIBRE3_CHAR_HISTORIC_DATA, but not LIBRE3_CHAR_GLUCOSE_DATA.

Why do you regularly cancel all your posts?

**Comment by RempageR1** [comment]
I`m running FSL3 with the patched app and Juggluco. Battery optimization etc is turned off, but when I get out of range and back in range, I have to turn off bluetooth and turn it on again to get it working. Anyone had this issue? Phone is a Samsung M22.

**Reply by maheini** [reply]
Do you have android version 13? There were a lot of Android 13 related bluetooth issues.

**Reply by RempageR1** [reply]
I had to check but yes, Android 13 it is. Do you have any suggestions on what to do? Or just wait for a new build of the FSL3 application?

**Reply by RempageR1** [reply]
In the end, I enabled "Bluetooth Wakelocks" in Xdrip which solved the problem. Hope that can help someone else with this issue as well.

**Reply by maheini** [reply]
Oh, those are great news. Very glad you found a solution which might help others

**Reply by RempageR1** [reply]
Thank you, yes I hope someone will read this and can perhaps confirm it works on their end as well.

**Comment by b3nene** [comment]
If you encounter the errorcode: "postgetauth: status=10 reason=invalidUserOrPassword" check your email-adress. It appears that if you use a old emailadress that started as a @googlemail.com you might have to use this address instead of your @gmail.com address. I have no answer to the why, I just encouterd this and wanted to share the solution I found.

**Reply by maheini** [reply]
Thank you for the hint 👍

**Comment by AlexandersPump** [comment]
Hello, I am using successfully Juggluco for more than a year to feed AAPS with glucose data from FSL3. Recently, I have changed the phone. Since then, I have issued strating a new sensor. I can seamlessly scan a new sensor, the apps is identifying teh sensor and counting down the warm-up time. However, after 60 minutes, Juggluco is NOT starting to receive data. When checking the settings (left menu --> sensor), I can see, that the sensor is identified, but it is stated "Streaming not enabled".  Of course BT is enabled. How can I activate streaming?

**Reply by AlexandersPump** [reply]
![IMG_0091](https://github.com/maheini/FreeStyle-Libre-3-patch/assets/51724637/cc8aac98-667e-4bf1-9e35-13b32f95b0d0)
it looks for me (not being an SW developper) that there is an issue with the BT permission for the Juggluco App; obvoiusly there has been a change re BT permissions in Android 12 and above; my new phone is running Android 13; just FYI: the Libre 3 app (on the same phone) from Abbott is able to read the sensor seamlessly ; of course killed when trying to connect to Juggluco

**Reply by j-kaltes** [reply]
Streaming is enabled by successfully scanning the sensor with „_Sensor via Bluetooth/Use Bluetooth_“ . Sometimes it is only shown under left menu→Sensor until after Juggluco successfully received glucose values from the sensor.
You should look for other reasons why you don’t receive glucose values from the sensor.

Bluetooth doesn’t seem to function. Turning Bluetooth off and on may help, otherwise rebooting the phone. Juggluco certainly makes use of the new Bluetooth permissions introduced in Android 12. On startup you are asked to give Juggluco “_Nearby devices_” permission. If you didn’t give Juggluco this permission, Bluetooth will not work, but you will be notified of that.
